### PR TITLE
split jsonrpc-types into jsonrpc-types and jsonrpc-proto

### PR DESCRIPTION
**Pull request by zeroqn**
jsonrpc-types only contains type definitions and some trivial functions.

### CODE REVIEW

File: jsonrpc-proto/src/complete.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+// This program is distributed in the hope that it will be
+// useful, but WITHOUT ANY WARRANTY; without even the implied
+// warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+// PURPOSE. See the GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use jsonrpc_types::{request::*, rpctypes::Params as PartialParams, Error};
```
Comment (chaoticlonghair) [inline jsonrpc-proto/src/complete.rs]: I / we hate `use xxx::*`.

File: jsonrpc-proto/src/request.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                    $(
+                        Call::$enum_name{ params } => params.try_into_proto(),
+                    )+
+                }
+            }
+        }
+    };
+}
+
+call_into_proto![
```
Comment (chaoticlonghair) [inline jsonrpc-proto/src/request.rs]: It become complex, after you split `define_call!()`.

We have to check if `partial_call_complete!()`, `call_into_proto!()` and `define_call!()` has same items.
Why don't put the trait into `jsonrpc-types`?

Comment (zeroqn) [reply jsonrpc-proto/src/request.rs]: Thanks for tips.
```partial_call_complete!()``` and ```call_into_proto!``` now use a macro from ```jsonrpc-types``` and
get types from there.

File: jsonrpc-types/src/rpctypes/basic/integer.rs [hunk 3]
```diff
@@ -84,11 +88,11 @@ macro_rules! impl_convert_with_uint {
                 Integer::new(data as u64)
             }
         }
+
         impl TryInto<$uint> for Integer {
```
Comment (chaoticlonghair) [inline jsonrpc-types/src/rpctypes/basic/integer.rs]: You do not have to add a trait.
IMHO, just impl a method is better: `fn try_into_uint<T> -> Result<T, Error>` or `fn try_into_uint<T> ->Option<T>`.
Do not let the old code to restrict u.

File: jsonrpc-types/src/request/request.rs [hunk 4]
```diff
@@ -275,29 +219,36 @@ pub trait JsonRpcRequest {
...
-    (GetFilterChanges, GetFilterChangesParams: [Quantity], FilterChanges),
-    (GetFilterLogs, GetFilterLogsParams: [Quantity], Logs),
-    (GetTransactionProof, GetTransactionProofParams: [Data32], Data),
-    (GetMetaData, GetMetaDataParams: [BlockNumber], MetaData),
-    (GetStateProof, GetStateProofParams: [Data20, Data32, BlockNumber], Data),
-    (GetBlockHeader, GetBlockHeaderParams: [BlockNumber], Data),
-    (GetStorageAt, GetStorageKeyParams: [Data20, Data32, BlockNumber], Data),
-);
+#[macro_export]
+macro_rules! loop_macro_for_types {
```
Comment (chaoticlonghair) [inline jsonrpc-types/src/request/request.rs]: I think, the name is not proper for `export`.
`construct_for_each_jsonrpc_types`, or else.

Comment (zeroqn) [reply jsonrpc-types/src/request/request.rs]: How about ```impl_for_each_jsonrpc_requests```?

File: jsonrpc-types/src/rpctypes/basic/integer.rs (hunk 3 above)
Comment (zeroqn) [reply jsonrpc-types/src/rpctypes/basic/integer.rs]: That trait is now shadowed.

**Review by chaoticlonghair — APPROVED** [review]

**Review by u2 — APPROVED** [review]

**Pull request merged**